interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/BaseNodeCompiler.scala

```diff
@@ -0,0 +1,325 @@
+}
+import pl.touk.nussknacker.engine.api.expression.{ExpressionTypingInfo, TypedExpression}
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectTypingResult, TypingResult, Unknown}
+import pl.touk.nussknacker.engine.compile._
+import pl.touk.nussknacker.engine.compile.nodecompilation.BaseNodeCompiler.combineErrors
+import pl.touk.nussknacker.engine.compile.nodecompilation.NodeCompiler.NodeCompilationResult
+import pl.touk.nussknacker.engine.compiledgraph.evaluatedparam.TypedParameter
+import pl.touk.nussknacker.engine.compiledgraph.variable
+import pl.touk.nussknacker.engine.graph.expression.NodeExpressionId.DefaultExpressionId
+import pl.touk.nussknacker.engine.graph.expression._
+import pl.touk.nussknacker.engine.graph.node
+import pl.touk.nussknacker.engine.graph.node._
+
+class BaseNodeCompiler(objectParametersExpressionCompiler: ExpressionCompiler) {
+
+  def compileVariable(variable: Variable, ctx: ValidationContext)(
+      implicit nodeId: NodeId
+  ): NodeCompilationResult[expression.Expression] = {
+    val (expressionCompilation, nodeCompilation) =
+      ExpressionCompilerAdapter.compileExpression(
+        variable.value,
+        ctx,
+        expectedType = Unknown,
+        outputVar = Some(OutputVar.variable(variable.varName))
+      )
+
+    val additionalValidationResult: ValidatedNel[ProcessCompilationError, Unit] =
+      ValidationAdapter.validateMaybeVariable(expressionCompilation.typedExpression, DefaultExpressionId)
+
+    combineErrors(nodeCompilation, additionalValidationResult)
+  }
+
+  def compileFilter(filter: Filter, ctx: ValidationContext)(
+      implicit nodeId: NodeId
+  ): NodeCompilationResult[expression.Expression] = {
+    val (expressionCompilation, nodeCompilation) =
+      ExpressionCompilerAdapter.compileExpression(
+        filter.expression,
+        ctx,
+        expectedType = Typed[Boolean],
+        outputVar = None
+      )
+
+    val additionalValidationResult =
+      ValidationAdapter.validateMaybeBoolean(expressionCompilation.typedExpression, DefaultExpressionId)
+
+    combineErrors(nodeCompilation, additionalValidationResult)
+  }
+
+  def compileSwitch(
+      expressionRaw: Option[(String, Expression)],
+      choices: List[(String, Expression)],
+      ctx: ValidationContext
+  )(
+      implicit nodeId: NodeId
+  ): NodeCompilationResult[(Option[expression.Expression], List[expression.Expression])] = {
+
+    // the frontend uses empty string to delete deprecated expression.
+    val expression = expressionRaw.filterNot(_._1.isEmpty)
+
+    val expressionCompilation = expression.map { case (output, expression) =>
+      ExpressionCompilerAdapter
+        .compileExpression(
+          expression,
+          ctx,
+          Unknown,
+          NodeExpressionId.DefaultExpressionId,
+          Some(OutputVar.switch(output))
+        )
+        ._2
+    }
+    val objExpression = expressionCompilation.map(_.compiledObject.map(Some(_))).getOrElse(Valid(None))
```

Comment:
It can be just `expressionCompilation.map(_.compiledObject).sequence`

Reply:
I'm still not used to cats it seems :D Done